src/main/java/bio/terra/service/common/QueryUtils.java

```diff
@@ -0,0 +1,22 @@
+package bio.terra.service.common;
+
+import bio.terra.grammar.Query;
+import bio.terra.grammar.exception.InvalidFilterException;
+import bio.terra.grammar.exception.InvalidQueryException;
+import org.apache.commons.lang3.StringUtils;
+
+public class QueryUtils {
+  public static String formatAndParseUserFilter(String filter) {
+    if (StringUtils.isEmpty(filter)) return "";
+    String filterWithWhere =
+        (filter.matches("(?i)\\(?where\\s+.*\\)?")) ? filter : "WHERE (" + filter + ")";
+    // Parse a Sql skeleton with the filter
+    // Note: We could specifically parse the where statement, but it doesn't catch as many errors
+    try {
+      Query.parse("select * from schema.table " + filterWithWhere);
```

Comment:
By adding the "select * from schema.table " to the filter clause, we're able to give the ANTLR code a full SQL statement for it to parse and make sure it's a grammatically correct filter clause. The ANTLR grammar doesn't know anything about the actual validity of the SQL statement, just that the words/parentheses/periods etc. are in the correct places so that it could be a valid SQL statement. So, "schema" and "table" do not relate back to anything. 

I actually first tried just parsing the "WHERE" clause. This is easy/possible with ANTLR, but I found that it wasn't catching a lot of errors. For example, it caught that "WHERE a = 1)" was invalid but not "WHERE (a = 1". (OR maybe vice-versa?, I don't remember) because it fumbled with end of line statements when it wasn't using the full grammar. I think I could have fixed this by beefing up the grammar for WHERE clauses, but I saw that @nmalfroy had actually done this skeleton grammar in another spot, so I borrowed that code!

Reply:
Ohhhh!  Thanks so much, I understand what you meant by "SQL skeleton" now.